land_grab_2/stl_dataset/step_2_5/get_cessions.py

```diff
@@ -0,0 +1,278 @@
+                              additional columns
+
+    """
+    for v in variables:
+        if v in target_df.columns:
+            raise ValueError(f"Column '{v}' already present in target_df.")
+
+    target_df = target_df.copy()
+    target_ix, source_ix = source_df.sindex.query(
+        target_df.geometry, predicate="intersects"
+    )
+    areas = (
+        target_df.geometry.values[target_ix]
+        .intersection(source_df.geometry.values[source_ix])
+        .area
+    )
+
+    main = []
+    for i in range(len(target_df)):
+        mask = target_ix == i
+        if np.any(mask):
+            main.append(source_ix[mask][np.argmax(areas[mask])])
+        else:
+            main.append(np.nan)
+
+    main = np.array(main, dtype=float)
+    mask = ~np.isnan(main)
+
+    for v in variables:
+        arr = np.empty(len(main), dtype=object)
+        arr[mask] = source_df[v].values[main[mask].astype(int)]
+        try:
+            arr = arr.astype(source_df[v].dtype)
+        except TypeError:
+            warnings.warn(
+                f"Cannot preserve dtype of '{v}'. Falling back to `dtype=object`.",
+            )
+        target_df[v] = arr
+
+    return target_df
+
+
+def join_counties_to_parcels(
+    parcels_gdf: gpd.GeoDataFrame, counties_gdf: gpd.GeoDataFrame
+) -> gpd.GeoDataFrame:
+    """
+    Join county name to parcels based on the largest spatial intersection.
+
+    Arguments:
+    parcels_gdf -- GeoDataFrame containing parcel geometries
+    counties_gdf -- GeoDataFrame containing county geometries
+
+    Returns:
+    GeoDataFrame -- STL parcels with county names joined as a new column
+    """
+    # Reproject counties_gdf to match parcels_gdf.
+    counties_gdf = counties_gdf.to_crs(parcels_gdf.crs)
+
+    out_gdf = area_join(
+        counties_gdf,
+        parcels_gdf,
+        variables=["NAME"],
+    )
+
+    # Drop the existing county column—we'll use the result of this join instead.
+    out_gdf = out_gdf.drop(columns=["county"])
+    out_gdf = out_gdf.rename(columns={"NAME": "county"})
+
+    return out_gdf
+
+
+def join_cessions_to_parcels(
+    parcels_gdf: gpd.GeoDataFrame, cessions_gdf: gpd.GeoDataFrame
+) -> gpd.GeoDataFrame:
+    """
+    Join parcel attributes to cessions based on all spatial intersections.
+
+    This is a one-to-many left join; a given parcel may intersect many (up to 8)
+    cessions.
+
+    Arguments:
+    parcels_gdf -- GeoDataFrame containing parcel geometries
+    cessions_gdf -- GeoDataFrame containing cession geometries
+
+    Returns:
+    GeoDataFrame -- STL parcels with each parcel-cession intersection as a dis-
+                    tinct row
+    """
+    # Reproject cessions_gdf to match parcels_gdf.
+    cessions_gdf = cessions_gdf.to_crs(parcels_gdf.crs)
+
+    # Subset cessions_gdf to only include the CESSNUM and geometry columns.
+    cessions_gdf = cessions_gdf[["CESSNUM", "geometry"]]
+
+    out_gdf = parcels_gdf.sjoin(cessions_gdf, how="left")
```

Comment:
As explained in the docstring, we use a standard one-to-many spatial join here (not area-weighted). I used a left join here to ensure that we didn't drop any parcels, but, looking at the eventual output, every parcel does have a matching cession. I think we should be safe with an inner join, but let me double check.